File: 백준/Gold/2470. 두 용액/두 용액.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
using namespace std;

int main() {
	cin.tie(NULL);
	ios::sync_with_stdio(false);


	int n;
	cin >> n;
	vector<int> v;
	int temp;
	for (int i = 0; i < n; i++) {
		cin >> temp;
		v.push_back(temp);
	}

	sort(v.begin(), v.end());
	int s = 0, e = n - 1;
	int sum = 0;
	int min = 2000000000;
	vector<int> ans(2);
	while (s < e) {
		sum = v[s] + v[e];
		if (abs(sum) < min) {
			min = abs(sum);
			ans[0] = v[s];
			ans[1] = v[e];
			if (sum == 0) break;
		}
		if (sum < 0) {
			s++;
		}
		else {
			e--;
		}
	}
	cout << ans[0] << " " << ans[1];
}

File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;

		int n = Integer.parseInt(br.readLine());
		int[] arr = new int[n];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		Arrays.sort(arr);
		
		int left = 0;
		int right = n - 1;
		int min = Integer.MAX_VALUE;
		int[] ans = new int[2];
		while(left < right) {
			int sum = arr[left] + arr[right];
			
			if(Math.abs(sum) < min) {
				min = Math.abs(sum);
				ans[0] = arr[left];
				ans[1] = arr[right];
			}
			
			if(sum < 0) {
				left++;
			}
			else right--;
		}
		
		System.out.println(ans[0] + " " + ans[1]);
	}
}
